### ISSUE TITLE
Multiple expanding advertisments can be dropped preventing devices from being discovered

### ISSUE THREAD

**Issue by blackspherefollower** [buttplugio/buttplug#462]
**Describe the bug**
The BTLE device scanner receives duplicate advertisements from devices, so rather than flood the device configuration layer with events, it only sends new device events when the advertisement has changed. In some cases, those advertisement changes can come in really fast, causing 2 events to be fired in quick succession.

The event receiver throws away new events for connected devices AND events for devices being considered. The later occasionally causes devices to be missed if a device event arrives with a non-matching advertisement followed too soon by an advertising that would match, but is dropped because the device is already been considered.

**Expected behavior**
Since the events can't return context to their senders (I think), I believe that rather than dropping events for devices being considered, that the event should be held until the existing consideration is complete, then if the device isn't connected, it should be considered again.

**Actual behavior**
The device is blocked by the scanner until the scan is restarted.

**Comment by qdot** [comment]
Hmm, this may still happen in v6. Honestly we may just want to consider devices *every* time we see their advertisement but instead of ignoring them completely, just log once that we're not doing anything with them then continuously throw advertisements away in the background? I'm not sure how much processing that would require tho.

**Comment by blackspherefollower** [comment]
Unfortuatly the fix committed did not resolve the issue:
```
2022-09-26T10:04:32.280104Z DEBUG btleplug enumeration{address=PeripheralId(A4:C1:38:64:E5:3F) name=}: buttplug::server::device::hardware::communication::btleplug::btleplug_adapter_task: Found new bluetooth device advertisement: PeripheralInfo { name: None, peripheral_id: PeripheralId(A4:C1:38:64:E5:3F), services: [0000180a-0000-1000-8000-00805f9b34fb] }
2022-09-26T10:04:32.280259Z  INFO buttplug::server::device::server_device_manager_event_loop: Device  (PeripheralId(A4:C1:38:64:E5:3F)) found.
2022-09-26T10:04:32.280292Z DEBUG buttplug::server::device::server_device_manager_event_loop: Device PeripheralId(A4:C1:38:64:E5:3F) allowed via configuration file, continuing.
2022-09-26T10:04:32.280361Z DEBUG buttplug::server::device::configuration: Looking for protocol that matches specifier: BluetoothLE(BluetoothLESpecifier { names: {""}, advertised_services: {0000180a-0000-1000-8000-00805f9b34fb}, services: {} })
2022-09-26T10:04:32.280511Z  INFO buttplug::server::device::configuration: Found protocol "satisfyer" for specifier BluetoothLE(BluetoothLESpecifier { names: {""}, advertised_services: {0000180a-0000-1000-8000-00805f9b34fb}, services: {} }).
2022-09-26T10:04:32.282348Z DEBUG btleplug enumeration{address=PeripheralId(A4:C1:38:64:E5:3F) name=bobi2}: buttplug::server::device::hardware::communication::btleplug::btleplug_adapter_task: Found new bluetooth device advertisement: PeripheralInfo { name: Some("bobi2"), peripheral_id: PeripheralId(A4:C1:38:64:E5:3F), services: [0000180a-0000-1000-8000-00805f9b34fb] }
2022-09-26T10:04:32.282481Z  INFO buttplug::server::device::server_device_manager_event_loop: Device bobi2 (PeripheralId(A4:C1:38:64:E5:3F)) found.
2022-09-26T10:04:32.282514Z DEBUG buttplug::server::device::server_device_manager_event_loop: Device PeripheralId(A4:C1:38:64:E5:3F) allowed via configuration file, continuing.
2022-09-26T10:04:32.282588Z DEBUG buttplug::server::device::configuration: Looking for protocol that matches specifier: BluetoothLE(BluetoothLESpecifier { names: {"bobi2"}, advertised_services: {0000180a-0000-1000-8000-00805f9b34fb}, services: {} })
2022-09-26T10:04:32.282729Z  INFO buttplug::server::device::configuration: Found protocol "magic-motion-4" for specifier BluetoothLE(BluetoothLESpecifier { names: {"bobi2"}, advertised_services: {0000180a-0000-1000-8000-00805f9b34fb}, services: {} }).
2022-09-26T10:04:32.282811Z  INFO buttplug::server::device::configuration: Found protocol "satisfyer" for specifier BluetoothLE(BluetoothLESpecifier { names: {"bobi2"}, advertised_services: {0000180a-0000-1000-8000-00805f9b34fb}, services: {} }).
2022-09-26T10:04:32.283050Z  INFO buttplug::server::device::server_device_manager_event_loop: Device PeripheralId(A4:C1:38:64:E5:3F) currently trying to connect, ignoring new device event.
2022-09-26T10:04:33.317491Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: btleplug::winrtble::ble::device: services 6    
2022-09-26T10:04:33.488233Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: btleplug::winrtble::ble::device: characteristics Ok(3)    
2022-09-26T10:04:33.614530Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-09-26T10:04:33.808933Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: btleplug::winrtble::ble::device: characteristics Ok(5)    
2022-09-26T10:04:33.879518Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-09-26T10:04:34.088854Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: btleplug::winrtble::ble::device: characteristics Ok(8)    
2022-09-26T10:04:34.147177Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: btleplug::winrtble::ble::device: characteristics Ok(1)    
2022-09-26T10:04:34.147975Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: buttplug::server::device::hardware::communication::btleplug::btleplug_hardware: Found required service 0000180a-0000-1000-8000-00805f9b34fb Service { uuid: 0000180a-0000-1000-8000-00805f9b34fb, primary: true, characteristics: {Characteristic { uuid: 00002a24-0000-1000-8000-00805f9b34fb, service_uuid: 0000180a-0000-1000-8000-00805f9b34fb, properties: READ }, Characteristic { uuid: 00002a25-0000-1000-8000-00805f9b34fb, service_uuid: 0000180a-0000-1000-8000-00805f9b34fb, properties: READ }, Characteristic { uuid: 00002a26-0000-1000-8000-00805f9b34fb, service_uuid: 0000180a-0000-1000-8000-00805f9b34fb, properties: READ }, Characteristic { uuid: 00002a27-0000-1000-8000-00805f9b34fb, service_uuid: 0000180a-0000-1000-8000-00805f9b34fb, properties: READ }, Characteristic { uuid: 00002a29-0000-1000-8000-00805f9b34fb, service_uuid: 0000180a-0000-1000-8000-00805f9b34fb, properties: READ }} }
2022-09-26T10:04:34.148068Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: buttplug::server::device::hardware::communication::btleplug::btleplug_hardware: Found characteristic 00002a24-0000-1000-8000-00805f9b34fb for endpoint rxblemodel
2022-09-26T10:04:34.172023Z  INFO device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: buttplug::server::device::protocol::satisfyer: Satisfyer Device Identifier: [1, 2, 36, 16] 16917520
2022-09-26T10:04:34.172130Z DEBUG device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: buttplug::server::device::configuration: Protocol device config found for ServerDeviceIdentifier { address: "PeripheralId(A4:C1:38:64:E5:3F)", protocol: "satisfyer", attributes_identifier: Identifier("16917520") }
2022-09-26T10:04:34.172263Z ERROR device creation{name= address=PeripheralId(A4:C1:38:64:E5:3F)}: buttplug::server::device::server_device_manager_event_loop: Device errored while trying to connect: Device does not have endpoint command
```

The log snippet above shows the MagicMotion Bobi's advertisements arrive in quick succession:
1. advertised_services: {0000180a-0000-1000-8000-00805f9b34fb}
2. names: {"bobi2"}

The first advertisement matches the Satisfyer protcol, causing connect and endpoint checks to fire off.
Whilst we're connecting and scanning the services, the second advertisement notification that adds the BLE name fires off, but gets ignored because we're already testing the device against the first advertisement info (without the BLE name).

Now at this point, we've recieved both a BLE name and a set of services, so the check in buttplug/src/server/device/hardware/communication/btleplug/btleplug_adapter_task.rs#94 skips raising any further advertisements for the device, so now the device has been ignored in the DeviceFound device manager event handler and will be ignored by the BtlePlug advertisement emitter, the Bobi will never be resolved.

**Comment by qdot** [comment]
Just need to switch over to checking for name AND address when we're checking dupes.

**Comment by qdot** [comment]
Implemented manufacturer data checking in v6.1.0, which should fix this for now because we have less of a chance of passing through incomplete advertisements to the connection step. This is definitely still a problem if we ever need shared advertised service checking again, but at this point I can't think of a reason we'd connect to something without having enough info. Closing for now, will reopen if this returns to being an issue.
